### PR TITLE
add v2-layer

### PR DESCRIPTION
**Pull request by jacquesqiao**
Before this new package `paddle.v2.layer`, users would need to use functions in `paddle.trainer_config_helpers.layers` to configure networks.

### The Old Way

This old way requires that the creation of a network be defined in a Python function, say `network_config`, and that this Python function being passed to `paddle.trainer_config_helpers.parse_network_config` for the creation of protobuf message description of this network.

```python
def network_config():
  img = paddle.trainer_config_helpers.data_layer(name="pixel", size=784)
  inference = paddle.trainer_config_helpers.fc_layer(
    input=img, 
    size=10, 
    act=paddle.trainer_config_helpers.SoftmaxActivation())
  cost = paddle.trainer_config_helpers.classification_cost(
    input=inference, 
    label=paddle.trainer_config_helpers.data_layer(name="label", size=10))
  
proto_desc = parse_network_config(network_config)
```

When `parse_network_config` executes `network_config`, those layer definition functions like `data_layer` and `fc_layer` would change some Python global variables, so that after the execution, `parse_network_config`  could collect information from these global variables and generates the protobuf message.

### The New Way

In this PR, we define a function in `paddle.v2.layer` which creates a Python class for each layer creation function in `paddle.trainer_config_helpers.layers`.  Users can use create a network as follows:

```python
img = paddle.v2.layer.data(name="pixel", size=784)
inference = paddle.v2.layer.fc(input=img, size=10, act=paddle.v2.layer.Softmax())
cost = paddle.v2.layer.classification(
  input=inference, 
  label=paddle.v2.layer.data(name="label", size=10))

parameters = paddle.v2.parameters.create(cost)
```

This new way doesn't require those invocations to layer definition functions to be in a Python function but could be anywhere.

Also, the creation of a protobuf message is hidden in the invocation of `paddle.v2.parameters.create`, no longer exposed to users.

### CODE REVIEW

File: python/paddle/v2/layer.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+# limitations under the License.
+
+import paddle.trainer_config_helpers as conf_helps
+from paddle.trainer_config_helpers.config_parser_utils import \
+    parse_network_config as __parse__
+from paddle.trainer_config_helpers.default_decorators import wrap_name_default
+import collections
+
+
+def parse_network(*outputs):
```
Comment (jacquesqiao) [inline python/paddle/v2/layer.py]: 因为是复用了之前的layer配置的代码，所以这里最终还是调用了paddle.trainer_config_helpers.config_parser_utils中的parse_network_config

File: python/paddle/v2/layer.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        raise NotImplementedError()
+
+
+def __convert_to_v2__(method_name, name_prefix, parent_names):
+    if name_prefix is not None:
+        wrapper = wrap_name_default(name_prefix=name_prefix)
+    else:
+        wrapper = None
+
+    class V2LayerImpl(Layer):
```
Comment (jacquesqiao) [inline python/paddle/v2/layer.py]: 这里是把原来定义layer的function 封装成了一个class，真正的调用是在
to_proto_impl中的getattr(conf_helps, method_name)(name=self.name, **args)

File: python/paddle/v2/layer.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+
+class Layer(object):
+    def __init__(self, name, parent_layer):
+        assert isinstance(parent_layer, dict)
+        assert isinstance(name, basestring)
+        self.name = name
+        self.__parent_layer__ = parent_layer
+
+    def to_proto(self, context):
```
Comment (jacquesqiao) [inline python/paddle/v2/layer.py]: context中保存了已经parse过的proto，如果自定义了name，而且name一样，这个layer就只会parse一次

File: demo/mnist/api_train.py [hunk 4]
```diff
@@ -67,7 +58,7 @@ def input_order_converter(generator):
 def main():
     api.initPaddle("-use_gpu=false", "-trainer_count=4")  # use 4 cpu cores
```
Comment (gangliao) [inline demo/mnist/api_train.py]: 这里好像包成了
```
paddle_v2.init(use_gpu=False, trainer_count=1)
```

Comment (gangliao) [inline demo/mnist/api_train.py]: https://github.com/PaddlePaddle/Paddle/pull/1324/files#diff-f1527307b4c2465cdb80925591bc7071R22

Comment (jacquesqiao) [inline demo/mnist/api_train.py]: 嗯，对，这两部分代码合并之后就改成上面这种方式的

File: demo/mnist/api_train.py [hunk 5]
```diff
@@ -79,8 +70,18 @@ def main():
...
     assert isinstance(updater, api.ParameterUpdater)
 
+    # define network
+    images = paddle_v2.layer.data(name='pixel', size=784)
+    label = paddle_v2.layer.data(name='label', size=10)
+    hidden1 = paddle_v2.layer.fc(input=images, size=200)
+    hidden2 = paddle_v2.layer.fc(input=hidden1, size=200)
+    inference = paddle_v2.layer.fc(input=hidden2,
+                                   size=10,
+                                   act=SoftmaxActivation())
```
Comment (gangliao) [inline demo/mnist/api_train.py]: 描述网络的时候，有个小细节，总感觉SoftmaxActivation()一类的激活函数，不需要Activation命名，感觉比较冗余。

**Comment by wangkuiyi** [comment]
这个PR的目的是为了不需要调用原来的output函数？

File: demo/mnist/api_train.py [hunk 6]
```diff
@@ -79,8 +70,18 @@ def main():
...
+    cost = paddle_v2.layer.classification_cost(input=inference, label=label)
+
     # Create Simple Gradient Machine.
-    model_config = parse_network_config(network_config)
+    model_config = paddle_v2.layer.parse_network(cost)
```
Comment (helinwang) [inline demo/mnist/api_train.py]: 请问需要暴露parse_network这个细节吗？
还是需要的时候在不让用户知道的情况下parse_network

File: demo/mnist/api_train.py (hunk 5 above)
Comment (helinwang) [inline demo/mnist/api_train.py]: 请问softmax是activation吗？我的理解是relu这些是activation。。。

Comment (gangliao) [inline demo/mnist/api_train.py]: ReLU， Sigmoid， Softmax， Linear应该都是activation吧

Comment (gangliao) [inline demo/mnist/api_train.py]: https://keras.io/activations/

File: demo/mnist/api_train.py (hunk 6 above)
Comment (jacquesqiao) [inline demo/mnist/api_train.py]: 不需要了，这个调用封装在paddle.parameters.create(topology)里边去，用户看不到

File: demo/mnist/api_train.py (hunk 5 above)
Comment (jacquesqiao) [inline demo/mnist/api_train.py]: done，已修改

**Comment by jacquesqiao** [comment]
已按照@gangliao意见修改了activation命名，在v2下引入了activation相关定义。

File: demo/mnist/api_train.py (hunk 5 above)
Comment (helinwang) [inline demo/mnist/api_train.py]: @gangliao 明白了，看来是我理解错了。谢谢！

**Comment by wangkuiyi** [comment]
I push a commit, which resolved the conflict with `upstream/develop`. I ran `demo/mnist/api_train.py`, it seems works.
